static/js/util/date_validation.js

```diff
@@ -0,0 +1,83 @@
+// @flow
+import R from 'ramda';
+import moment from 'moment';
+
+import { S, ifNil } from './sanctuary';
+const { Maybe, Nothing } = S;
+import { filterPositiveInt } from './util';
+import { YEAR_VALIDATION_CUTOFF } from '../constants';
+
+/**
+ * Removes non-numeric characters and truncates output string
+ */
+export const sanitizeNumberString = R.curry((length: number, input: string|number): string => {
+  if ( typeof input === 'string' ) {
+    let out = input.replace(/[^\d]+/g, '');
+    if ( out.match(/^0+/) ) {
+      if ( out.length <= length ) {
+        return out.slice(0, length);
+      } else {
+        return out.replace(/^0+/, "").slice(0, length);
+      }
+    } else {
+      return out.slice(0, length);
+    }
+  } else {
+    return String(input).slice(0, length);
+  }
+});
+
+/**
+  * validate a day input
+  */
+export const checkDayRange = ifNil(day => day > 31 ? Maybe.of(31) : Maybe.of(day));
+
+export const validateDay = R.compose(
+  checkDayRange, filterPositiveInt, sanitizeNumberString(2)
+);
+
+/**
+ * Validate a month input
+ */
+export const checkMonthRange = ifNil(month => month > 12 ? Maybe.of(12) : Maybe.of(month));
+
+export const validateMonth = R.compose(
+  checkMonthRange, filterPositiveInt, sanitizeNumberString(2)
+);
+
+/**
+  * Validate a year input
+  * 
+  * checks to make sure a year n is in the range of
+  * x - YEAR_VALIDATION_CUTOFF <= n <= highCutoff,
+  * where x is moment().year()
+  */
+export const checkYearRange = R.curry((highCutoff, year) => {
+  if (year === undefined) {
+    return Nothing();
+  }
+  let now = moment().year();
+  let lowCutoff = now - YEAR_VALIDATION_CUTOFF;
+  if ( year < lowCutoff ) {
+    if ( String(year).length < 4 ) {
+      return Maybe.of(year);
+    }
+    return Maybe.of(lowCutoff);
+  }
+  if ( year >= highCutoff ) {
+    return Maybe.of(highCutoff);
+  }
+  return Maybe.of(year);
+});
+
+
+export const validateYear = ifNil(
+  R.compose(checkYearRange(moment().year()), filterPositiveInt, sanitizeNumberString(4))
+);
+
+/**
+  * checks that year is valid, with years up to 10 years in the future being allowable
```

Comment:
I'm a bit uneasy having any upper limit for graduation date, though it seems like a sensible one